### DISCUSSION TITLE
How to include computed properties in model dump?

### DISCUSSION THREAD

**Discussion by berzi** [pydantic/pydantic#7558] [Question] [answered]
Given the following example:

```python
from pydantic import BaseModel


class SubModel(BaseModel):
    txt: str = "test"

    @property
    def formatted_txt(self):
        return self.txt.title()

    def model_dump(self, **kwargs):
        return {
            **super().model_dump(**kwargs),
            "formatted_txt": self.formatted_txt,
        }


class MyModel(BaseModel):
    sub: SubModel = SubModel()


assert "formatted_txt" in SubModel().model_dump()  # OK
assert "sub" in MyModel().model_dump()  # OK
assert "formatted_txt" in MyModel().model_dump()["sub"]  # AssertionError
```

From the docs for `model_dump()`, emphasis mine:

> This is the primary way of converting a model to a dictionary. Sub-models will be **recursively converted** to dictionaries.

I assumed that meant that `model_dump()` would be called for sub-models recursively, but that doesn't seem to be the case.

This is counter-intuitive to me, but I'm aware of `@model_serializer()`, so I tried to implement what I wanted that way. But if I do the following:

```python
class SubModel(BaseModel):
    txt: str = "test"

    @property
    def formatted_txt(self):
        return self.txt.title()

   @model_serializer
   def ser_model(self) -> dict[str, Any]:
       return {
           "active": self.active,
       }
```

Then the fields defined on the model normally (here `txt`) will not appear at all. But I can't just include the normal `model_dump()` in `ser_model()`'s return value because the following:

```python
class SubModel(BaseModel):
    txt: str = "test"

    @property
    def formatted_txt(self):
        return self.txt.title()

   @model_serializer
   def ser_model(self) -> dict[str, Any]:
       return {
           **self.model_dump(),
           "active": self.active,
       }
```

...causes a RecursionError.

**So what is the correct way to include additional fields in the serialisation without getting rid of the default behaviour?**

And as a note: I think it would be less surprising and generally easier if `model_dump()` were called recursively by Pydantic.

**Comment by Viicos** [comment, marked as answer]
Take a look at [`computed_field`](https://docs.pydantic.dev/latest/usage/computed_fields/), designed for your use case

**Reply by berzi** [reply]
Thanks, I totally missed that. I'll see if I can add a link in the docs to help people like me. :')

**Reply by aviadr1** [reply]
@Viicos how can I also convince pydantic to include this computed field in the schema?

**Reply by Viicos** [reply]
It will be included by default in the JSON Schema. Do you have an example where it isn't the case?

**Reply by aviadr1** [reply]
example:
```python
from pydantic import BaseModel, Field, computed_field

class MyModel(BaseModel):
    first_name: str
    last_name: str

    @computed_field
    @property
    def full_name(self) -> str:
        """A computed property that concatenates first and last names."""
        return f"{self.first_name} {self.last_name}"

# Example usage
my_instance = MyModel(first_name='John', last_name='Doe')
print(json.dumps(my_instance.model_json_schema(), indent=2))

print()

print(my_instance.model_dump_json())
```
schema output (no full_name:
```
{
  "properties": {
    "first_name": {
      "title": "First Name",
      "type": "string"
    },
    "last_name": {
      "title": "Last Name",
      "type": "string"
    }
  },
  "required": [
    "first_name",
    "last_name"
  ],
  "title": "MyModel",
  "type": "object"
}
```


JSON serializtion (has full_name):
```
{"first_name":"John","last_name":"Doe","full_name":"John Doe"}
```

**Reply by brianmedigate** [reply]
try `MyModel.model_json_schema(mode="serialization")`

https://docs.pydantic.dev/latest/concepts/json_schema/#configuring-the-jsonschemamode

**Comment by brianmedigate** [comment]
In case anyone found this discussion trying to make a recursive (or "super()") call to `model_dump()` for other reasons which aren't the use case of `computed_field`, the way to do that is with the "wrap" mode of `@model_serializer()`.

The docs are a bit terse on the subject.

For example, to rewrite the [example from the documentation](https://docs.pydantic.dev/latest/api/functional_serializers/#pydantic.functional_serializers.model_serializer) in a way that includes any other fields the model has without having to duplicate them in the serializer:

```python
from datetime import date
from typing import Literal

from pydantic import BaseModel, model_serializer

class TemperatureModel(BaseModel):
    unit: Literal['C', 'F']
    value: int
    recorded_on: date

    @model_serializer(mode="wrap")
    def serialize_model(self, handler):
        result = handler(self)
        if self.unit == 'F':
            result.update({'unit': 'C', 'value': int((self.value - 32) / 1.8)})
        return result

temperature = TemperatureModel(unit='F', value=212, recorded_on=date(2024, 1, 1))
print(temperature.model_dump())
#> {'unit': 'C', 'value': 100, 'recorded_on': datetime.date(2024, 1, 1)}
```

**Reply by royp-nrd** [reply]
wondering how I can add additional runtime arg to model_dump? 

I'm trying to create this behavior:
```python
class TemperatureModel(BaseModel):
    unit: Literal['C', 'F']
    value: int
    recorded_on: date
   _a: str | None = PrivateAttr(None)

    @model_serializer(mode="wrap")
    def serialize_model(self, handler, info):
        result = handler(self)
        if info.include_private_attrs:
            result.update((self.__pydantic_private__ or {}))
        return result
temperature = TemperatureModel(unit='F', value=212, recorded_on=date(2024, 1, 1), _a="private")
print(temperature.model_dump(include_private_attrs=True))
#> {'unit': 'C', 'value': 100, 'recorded_on': datetime.date(2024, 1, 1), '_a': "private"}
```

Any idea?

**Reply by Viicos** [reply]
You can provide context in the `model_dump_*` methods, then access it via `info.context`.
